File: internal/headers/challenge.go
package headers

import (
	"net/http"
	"net/url"
	"strings"
)

// ChallengeHost is the exact Cloudflare-Turnstile widget host. Case-insensitive
// equality. Mirrors zp_shared::CHALLENGE_HOST; parity is enforced via
// crates/zp-shared/testdata/challenge_cases.json consumed by both
// `go test ./internal/headers` and `cargo test -p zp-shared`.
const ChallengeHost = "challenges.cloudflare.com"

// ChallengePlatformPrefix is the same-zone challenge-platform orchestration
// loader path prefix. Exact-prefix match on the URL path component (query and
// fragment never participate). Mirrors zp_shared::CHALLENGE_PLATFORM_PREFIX.
const ChallengePlatformPrefix = "/cdn-cgi/challenge-platform/"

// IsChallengeDocument is the pure predicate form. Returns true when:
//  1. cfMitigated, trimmed and lowered, equals "challenge"; or
//  2. host equals ChallengeHost case-insensitively; or
//  3. path starts with ChallengePlatformPrefix (path match is case-sensitive
//     because the upstream emits a fixed canonical lowercase path).
//
// Pure: no I/O, no body access. The Rust parity test in zp_shared::challenge
// shares the same golden vectors.
func IsChallengeDocument(cfMitigated, host, path string) bool {
	cf := strings.TrimSpace(cfMitigated)
	if cf != "" && strings.EqualFold(cf, "challenge") {
		return true
	}
	if strings.EqualFold(host, ChallengeHost) {
		return true
	}
	return strings.HasPrefix(path, ChallengePlatformPrefix)
}

// TargetIsChallengeDocument is the HTTP-typed wrapper around
// IsChallengeDocument. It reads ONLY headers and the final URL — never the
// response body. nil-safe on both args.
func TargetIsChallengeDocument(header http.Header, finalURL *url.URL) bool {
	cf := ""
	if header != nil {
		cf = header.Get("Cf-Mitigated")
	}
	host, path := "", ""
	if finalURL != nil {
		host = finalURL.Hostname()
		path = finalURL.Path
	}
	return IsChallengeDocument(cf, host, path)
}

// ChallengeSubresourceSkip is the two-signal gate for the ConstructorPolicy
// Cache-Control: no-store skip. Returns true ONLY when ALL hold: the tab is
// armed (operator opt-in), the request is NOT a document navigation
// (isDoc==false), and the response classifies as a challenge by header/URL.
// The isDoc==false term keeps the challenge DOCUMENT (navigation HTML) on
// no-store and lets ONLY classified SUBRESOURCES preserve Cloudflare's cache
// semantics for its turnstile api.js etc.
//
// Pure predicate, header+URL only; it never reads the body, grants no egress,
// and manufactures no eval.
func ChallengeSubresourceSkip(armed, isDoc bool, header http.Header, finalURL *url.URL) bool {
	if !armed || isDoc {
		return false
	}
	return TargetIsChallengeDocument(header, finalURL)
}

// ApplyChallengeCompat emits the internal X-ZP-Challenge-Compat marker header
// ONLY when BOTH gate signals are present: the per-tab arm opt-in (armed) AND
// header/URL classification as a challenge document. Inert otherwise, so the
// non-armed (default) path is byte-identical to today.
//
// B4 STRIP OBLIGATION: this internal marker MUST be consumed-and-deleted at
// the service-worker layer before the response reaches the proxied page, in
// the SAME place X-ZP-Dynamic-Compile is read+deleted. Without that strip the
// page sees the internal control header and the leak undermines the
// fingerprint-blind contract.
func ApplyChallengeCompat(header http.Header, armed bool, finalURL *url.URL) {
	if !armed || header == nil {
		return
	}
	if !TargetIsChallengeDocument(header, finalURL) {
		return
	}
	header.Set("X-ZP-Challenge-Compat", "1")
}

File: crates/zp-shared/src/challenge.rs
//! Cloudflare challenge classifier — pure header/URL predicate.
//!
//! Single source of truth shared with the Go server (`internal/headers/
//! challenge.go`); parity is enforced via `testdata/challenge_cases.json`
//! consumed by both `cargo test -p zp-shared` and `go test ./internal/headers`.
//!
//! Body is NEVER read or sniffed. Inputs are limited to three already-parsed
//! signals (the `Cf-Mitigated` response-header value, the final URL host, and
//! the final URL path). The classifier itself grants no egress, manufactures
//! no eval, and synthesizes no challenge solution — it is the gate that
//! upstream callers consult before applying the armed CSP / `no-store` skip /
//! `X-ZP-Challenge-Compat` marker.

/// The exact Cloudflare-Turnstile widget host. Case-insensitive equality.
pub const CHALLENGE_HOST: &str = "challenges.cloudflare.com";

/// The same-zone challenge-platform orchestration loader path prefix.
/// Exact-prefix match on the *path component* (caller must pre-extract it from
/// the URL, so query/fragment never participate).
pub const CHALLENGE_PLATFORM_PREFIX: &str = "/cdn-cgi/challenge-platform/";

/// Returns true when the response classifies as a Cloudflare challenge
/// document or subresource via:
///   1. `Cf-Mitigated: challenge` response header (case-insensitive, trimmed), OR
///   2. final-URL host == `challenges.cloudflare.com` (case-insensitive), OR
///   3. final-URL path starts with `/cdn-cgi/challenge-platform/`.
///
/// Pure: no I/O, no body access, no allocation beyond ASCII trimming.
pub fn is_challenge_document(cf_mitigated: &str, host: &str, path: &str) -> bool {
    let cf = cf_mitigated.trim();
    if !cf.is_empty() && cf.eq_ignore_ascii_case("challenge") {
        return true;
    }
    if host.eq_ignore_ascii_case(CHALLENGE_HOST) {
        return true;
    }
    path.starts_with(CHALLENGE_PLATFORM_PREFIX)
}

/// The two-signal SUBRESOURCE gate: returns true ONLY when the tab is armed
/// AND the request is not a document navigation AND the response classifies
/// as a challenge. The `is_doc == false` term is load-bearing: it keeps the
/// challenge DOCUMENT (navigation HTML) on `no-store` and lets ONLY classified
/// subresources (e.g. turnstile api.js) preserve Cloudflare's cache semantics.
pub fn challenge_subresource_skip(
    armed: bool,
    is_doc: bool,
    cf_mitigated: &str,
    host: &str,
    path: &str,
) -> bool {
    if !armed || is_doc {
        return false;
    }
    is_challenge_document(cf_mitigated, host, path)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    #[test]
    fn empty_inputs_are_not_a_challenge() {
        assert!(!is_challenge_document("", "", ""));
    }

    #[test]
    fn cf_mitigated_header_matches_case_insensitive_and_trimmed() {
        assert!(is_challenge_document("challenge", "example.test", "/x"));
        assert!(is_challenge_document("  Challenge  ", "example.test", "/x"));
        assert!(is_challenge_document("CHALLENGE", "example.test", "/x"));
        assert!(!is_challenge_document("managed", "example.test", "/x"));
        assert!(!is_challenge_document("", "example.test", "/x"));
    }

    #[test]
    fn challenge_host_matches_case_insensitive() {
        assert!(is_challenge_document("", "challenges.cloudflare.com", "/"));
        assert!(is_challenge_document("", "CHALLENGES.CLOUDFLARE.COM", "/"));
        assert!(!is_challenge_document(
            "",
            "challenges.cloudflare.com.evil.test",
            "/"
        ));
        assert!(!is_challenge_document(
            "",
            "evil.challenges.cloudflare.com",
            "/"
        ));
    }

    #[test]
    fn challenge_platform_prefix_matches_path_exactly() {
        assert!(is_challenge_document(
            "",
            "example.test",
            "/cdn-cgi/challenge-platform/h/g/orchestrate"
        ));
        assert!(!is_challenge_document(
            "",
            "example.test",
            "/cdn-cgi/challenge-platform"
        )); // missing trailing slash
        assert!(!is_challenge_document(
            "",
            "example.test",
            "/CDN-CGI/CHALLENGE-PLATFORM/"
        )); // path match is case-sensitive
        assert!(!is_challenge_document(
            "",
            "example.test",
            "/x/cdn-cgi/challenge-platform/"
        ));
    }

    #[test]
    fn subresource_gate_requires_armed_and_not_doc_and_classifier() {
        // ALL FOUR must hold: armed, !is_doc, classifier match.
        assert!(challenge_subresource_skip(
            true,
            false,
            "challenge",
            "x",
            "/"
        ));
        // disarmed → never
        assert!(!challenge_subresource_skip(
            false,
            false,
            "challenge",
            "x",
            "/"
        ));
        // is_doc → never (document stays on no-store)
        assert!(!challenge_subresource_skip(
            true,
            true,
            "challenge",
            "x",
            "/"
        ));
        // classifier miss → never
        assert!(!challenge_subresource_skip(true, false, "", "x", "/"));
    }

    #[derive(Deserialize)]
    struct Case {
        name: String,
        cf_mitigated: String,
        host: String,
        path: String,
        want: bool,
    }

    #[test]
    fn classifier_matches_golden_cases() {
        let raw = include_str!("../testdata/challenge_cases.json");
        let cases: Vec<Case> = serde_json::from_str(raw).expect("parse challenge_cases.json");
        assert!(!cases.is_empty(), "challenge_cases.json must not be empty");
        for c in cases {
            let got = is_challenge_document(&c.cf_mitigated, &c.host, &c.path);
            assert_eq!(
                got, c.want,
                "case {:?}: cf={:?} host={:?} path={:?} want {} got {}",
                c.name, c.cf_mitigated, c.host, c.path, c.want, got
            );
        }
    }
}
